File: src/main/java/no/fintlabs/access/control/service/FieldAccessService.java
package no.fintlabs.access.control.service;

import no.fintlabs.access.control.model.dto.FieldAccess;
import no.fintlabs.access.control.model.entity.FieldAccessEntity;
import no.fintlabs.access.control.model.entity.ResourceAccessEntity;
import no.fintlabs.access.control.model.metamodel.Metamodel;
import no.fintlabs.access.control.repository.FieldAccessEntityRepository;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.util.*;

@Service
public class FieldAccessService {

    private final FieldAccessEntityRepository fieldAccessEntityRepository;

    public FieldAccessService(FieldAccessEntityRepository fieldAccessEntityRepository) {
        this.fieldAccessEntityRepository = fieldAccessEntityRepository;
    }

    public List<FieldAccess> getFieldAccess(Metamodel metamodel, Optional<ResourceAccessEntity> resourceAccessEntity) {

        List<FieldAccess> result = new ArrayList<>(metamodel.fields().size());
        Collection<FieldAccessEntity> fieldAccessEntities = getFieldAccessEntities(resourceAccessEntity);

        for (String field : metamodel.fields()) {
            Optional<FieldAccessEntity> fieldAccessEntity = getFieldAccessEntity(field, fieldAccessEntities);

            result.add(
                    new FieldAccess(
                            field,
                            shouldContainConverter(fieldAccessEntity),
                            getIsHidden(fieldAccessEntity)
                    )
            );
        }

        return result;
    }

    private Collection<FieldAccessEntity> getFieldAccessEntities(Optional<ResourceAccessEntity> resourceAccessEntity) {
        if (resourceAccessEntity.isEmpty()) return Collections.emptyList();
        return fieldAccessEntityRepository.findByResourceId(resourceAccessEntity.get().getId());
    }

    private boolean getIsHidden(Optional<FieldAccessEntity> fieldAccessEntity) {
        if (fieldAccessEntity.isEmpty()) return false;
        return !fieldAccessEntity.get().getHasAccess();
    }

    private List<String> shouldContainConverter(Optional<FieldAccessEntity> fieldAccessEntity) {
        String input = fieldAccessEntity.isEmpty() ? "" : fieldAccessEntity.get().getMustContain();

        if (!StringUtils.hasText(input)) return Collections.emptyList();
        return Arrays.asList(input.split(","));
    }

    private Optional<FieldAccessEntity> getFieldAccessEntity(String field, Collection<FieldAccessEntity> fieldAccessEntities) {
        return fieldAccessEntities.stream()
                .filter(f -> f.getFieldName().equalsIgnoreCase(field))
                .findFirst();
    }
}

File: src/main/java/no/fintlabs/access/control/service/FieldAccessService.kt
package no.fintlabs.access.control.service

import no.fintlabs.access.control.model.dto.FieldAccess
import no.fintlabs.access.control.model.entity.FieldAccessEntity
import no.fintlabs.access.control.model.entity.ResourceAccessEntity
import no.fintlabs.access.control.model.metamodel.Metamodel
import no.fintlabs.access.control.repository.FieldAccessEntityRepository
import org.springframework.stereotype.Service

@Service
class FieldAccessService(private val fieldAccessEntityRepository: FieldAccessEntityRepository) {
    fun getFieldAccess(metamodel: Metamodel, resourceAccessEntity: ResourceAccessEntity?): List<FieldAccess> {
        val result: MutableList<FieldAccess> = ArrayList(metamodel.fields.size)
        val fieldAccessEntities = getFieldAccessEntities(resourceAccessEntity)

        for (field in metamodel.fields) {
            val fieldAccessEntity = getFieldAccessEntity(field, fieldAccessEntities)

            result.add(
                FieldAccess(
                    field,
                    shouldContainConverter(fieldAccessEntity),
                    getIsHidden(fieldAccessEntity)
                )
            )
        }

        return result
    }

    private fun getFieldAccessEntities(resourceAccessEntity: ResourceAccessEntity?): Collection<FieldAccessEntity> {
        val resourceId = resourceAccessEntity?.id ?: return emptyList()
        return fieldAccessEntityRepository.findByResourceId(resourceId)
    }

    private fun getIsHidden(fieldAccessEntity: FieldAccessEntity?): Boolean {
        return fieldAccessEntity?.hasAccess?.not() ?: false
    }

    private fun shouldContainConverter(fieldAccessEntity: FieldAccessEntity?): List<String> {
        val input = fieldAccessEntity?.mustContain?.takeIf { it.isNotBlank() } ?: return emptyList()
        return input.split(",").map { it.trim() }
    }

    private fun getFieldAccessEntity(
        field: String,
        fieldAccessEntities: Collection<FieldAccessEntity?>
    ): FieldAccessEntity? {
        return fieldAccessEntities.firstOrNull { it?.fieldName.equals(field, ignoreCase = true) }
    }
}
